library/billing/src/HCFA_1500.php

```diff
@@ -0,0 +1,771 @@
+    {
+        $this->hcfa_curr_line = 1;
+        $this->hcfa_curr_col = 1;
+        $this->hcfa_data = '';
+        $this->hcfa_proc_index = 0;
+    }
+
+    /**
+     * take the data element and place it at the correct coordinates on the page
+     *
+     * @global int $hcfa_curr_line
+     * @global int $hcfa_curr_col
+     * @global type $hcfa_data
+     * @param type $line
+     * @param type $col
+     * @param type $maxlen
+     * @param type $data
+     * @param type $strip   regular expression for what to strip from the data. period and has are the defaults
+     *                      02/12 version needs to include periods in the diagnoses hence the need to override
+     */
+    public function put_hcfa($line, $col, $maxlen, $data, $strip = '/[.#]/')
+    {
+        if ($line < $this->hcfa_curr_line) {
+            die("Data item at ($line, $col) precedes current line.");
+        }
+
+        while ($this->hcfa_curr_line < $line) {
+            $this->hcfa_data .= "\n";
+            ++$this->hcfa_curr_line;
+            $this->hcfa_curr_col = 1;
+        }
+
+        if ($col < $this->hcfa_curr_col) {
+            die("Data item at ($line, $col) precedes current column.");
+        }
+
+        while ($this->hcfa_curr_col < $col) {
+            $this->hcfa_data .= " ";
+            ++$this->hcfa_curr_col;
+        }
+
+        $data = preg_replace($strip, '', strtoupper($data));
+        $len = min(strlen($data), $maxlen);
+        $this->hcfa_data .= substr($data, 0, $len);
+        $this->hcfa_curr_col += $len;
+    }
+
+    /**
+     * Process the diagnoses for a given claim. log any errors
+     *
+     * @param type $claim
+     * @param string $log
+     */
+    private function process_diagnoses_02_12(&$claim, &$log)
+    {
+        $hcfa_entries=array();
+        $diags = $claim->diagArray(false);
+        if ($claim->diagtype=='ICD10') {
+            $icd_indicator='0';
+        } else {
+            $icd_indicator='9';
+        }
+
+        $hcfa_entries[]=new HCFA_Info(37, 42, 1, $icd_indicator);
+
+        // Box 22. Medicaid Resubmission Code and Original Ref. No.
+        $hcfa_entries[]=new HCFA_Info(38, 50, 10, $claim->medicaidResubmissionCode());
+        $hcfa_entries[]=new HCFA_Info(38, 62, 15, $claim->medicaidOriginalReference());
+
+        // Box 23. Prior Authorization Number
+        $hcfa_entries[]=new HCFA_Info(40, 50, 28, $claim->priorAuth());
+
+        $diag_count=0;
+        foreach ($diags as $diag) {
+            if ($diag_count<12) {
+                $this->add_diagnosis($hcfa_entries, $diag_count, $diag);
+            } else {
+                $log.= "***Too many diagnoses ".($diag_count+1).":".$diag;
+            }
+
+            $diag_count++;
+        }
+        // Sort the entries to put them in the page base sequence.
+        usort($hcfa_entries, array(self::cmp_hcfa_info()));
+
+        foreach ($hcfa_entries as $hcfa_entry) {
+            $hcfa_entry->put();
```

Comment:
This also hurts my brain

Reply:
(since that functions comes back to call the put_hfca function in this class...)

Reply:
ok, this may sound crazy, but perhaps we can pass this class instance to this function:
```
$hcfa_entry->put($this);
```